### PR TITLE
ExtFramebuffer cleanup

### PR DESCRIPTION
**Pull request by andr2000**
Hi,
please consider the below patches which add some more cleanup for ExtFramebuffer.
Without this patches you can get out of memory in use-cases which allocate/release
lots of PRIME buffers.

Thank you,
Oleksandr Andrushchenko

### CODE REVIEW

File: kms++/src/extframebuffer.cpp [hunk 1]
```diff
@@ -82,6 +83,25 @@ ExtFramebuffer::ExtFramebuffer(Card& card, uint32_t width, uint32_t height, Pixe
...
+			/*
+			 * do not throw an exception now, DRM core will
+			 * make proper cleanup, but still print an error
+			 */
+			if (ret)
+				printf("Error closing GEM %d (%s)\n", ret, strerror(errno));
+		}
+	}
+
 }
```
Comment (tomba) [inline kms++/src/extframebuffer.cpp]: I don't this is correct. Then again, it's not correct even without your patch...

The DRM buffers work in slightly mysterious ways: drmPrimeFDToHandle() does not always return a new handle. If the card fd already "knows" the buffer in question, drmPrimeFDToHandle() returns that same known handle.

So, let's say you have an application that creates a GEM buffer and frees it at exit. Between the alloc and free, it gives the buffer out as dmabuf fd to some other application. That other application gives the dmabuf fd back to our app, which does drmPrimeFDToHandle(). drmPrimeFDToHandle() returns the original handle. With your patch, this means that we'll free the same handle twice: once in ExtFramebuffer and then at exit time.

I think kms++ should have a hash table in the card class to refcount the handles. Although, how to handle platform specific buffers, then, like omap_bos...

File: kms++/src/extframebuffer.cpp [hunk 2]
```diff
@@ -87,6 +87,10 @@ ExtFramebuffer::~ExtFramebuffer()
 	for (unsigned i = 0; i < m_num_planes; ++i) {
 		FramebufferPlane& plane = m_planes[i];
 
+		/* unmap buffer */
+		if (plane.map)
+			munmap(plane.map, plane.size);
+
 		if (plane.prime_fd && plane.handle) {
```
Comment (tomba) [inline kms++/src/extframebuffer.cpp]: This one looks fine, but depends on the previous patch, so I can't cherry-pick it.

File: kms++/src/extframebuffer.cpp (hunk 1 above)
Comment (andr2000) [reply kms++/src/extframebuffer.cpp]: Well, I believe it is not a bug as all objects inside DRM are ref counted, e.g. when you do FDToHandle the corresponding refcount is incremented and user-space is expected to close the handle with GEM_CLOSE as well as FD needs a close on it.
Please check https://elixir.bootlin.com/linux/v4.16-rc2/source/drivers/gpu/drm/drm_prime.c#L39
on DMA-BUF/GEM Object references and lifetime
Actually, the reason for this patch was that without it Dumb buffers were only freed on application exit, so I was not able to recycle buffers at run-time

Comment (tomba) [reply kms++/src/extframebuffer.cpp]: I was referring to the "Self-importing" case in the comment you linked. Granted, I have tested it only on omapdrm, but I also discussed this on #dri-devel and the message was clear: you need a hash table (or similar) to track whether you have already "seen" the handle. This is why many of the drm libs have such a table.

I fully agree that the current implementation leaks. But I believe your fix can cause the issue I described.

Unfortunately I don't have a test case for this. It should not be difficult to write, though.

Comment (andr2000) [reply kms++/src/extframebuffer.cpp]: Well, you could try my self-import ;)
https://github.com/andr2000/kmsxx/tree/pr_add_ion_support/py/test_fwk
Just tested with self-import on R-Car M3 (kernel 4.9) and I see that
`Error closing GEM -1 (Invalid argument)`
`Error closing GEM -1 (Invalid argument)`
`[ 2170.006025] [drm:drm_ioctl] pid=1280, dev=0xe200, auth=1, DRM_IOCTL_GEM_CLOSE`
`[ 2170.006096] [drm:drm_ioctl] ret = -22`
Anyways, thanks

**Comment by tomba** [comment]
This one has been open for a long time, so closing. Feel free to send an updated series, or the unmap patch separately.

**Pull request closed without merging**